uint8[] matrix_links 
float32[] matrix_values
